File: app/src/main/java/com/hjc/base/viewmodel/LoadSirViewModel.java
package com.hjc.base.viewmodel;


import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;

import com.hjc.base.bean.ArticleBean;
import com.hjc.base.http.observer.BasePageObserver;
import com.hjc.base.model.LoadSirModel;
import com.hjc.baselib.viewmodel.BaseViewModel;

import java.util.List;

public class LoadSirViewModel extends BaseViewModel {

    private final LoadSirModel loadSirModel;

    private final MutableLiveData<List<ArticleBean.DataBean.DatasBean>> listLiveData = new MutableLiveData<>();

    private final MutableLiveData<Boolean> refreshData = new MutableLiveData<>();

    public LoadSirViewModel(@NonNull Application application) {
        super(application);
        loadSirModel = new LoadSirModel();
    }

    public void loadList(int page, boolean isFirst) {
        loadSirModel.load(page).subscribe(new BasePageObserver<ArticleBean>(this, isFirst) {
            @Override
            public void onSuccess(ArticleBean result) {
                refreshData.setValue(true);
                ArticleBean.DataBean dataBean = result.getData();
                if (dataBean != null) {
                    List<ArticleBean.DataBean.DatasBean> dataList = dataBean.getDatas();
                    if (dataList != null && dataList.size() > 0) {
                        listLiveData.setValue(dataList);
                    }
                }
            }

            @Override
            protected void onFailure(@NonNull Throwable e) {
                super.onFailure(e);
                refreshData.setValue(true);
                showError();
            }
        });
    }

    public MutableLiveData<List<ArticleBean.DataBean.DatasBean>> getListLiveData() {
        return listLiveData;
    }

    public MutableLiveData<Boolean> getRefreshData() {
        return refreshData;
    }
}

File: module-frame/src/main/java/com/hjc/module_frame/viewmodel/LoadSirViewModel.kt
package com.hjc.module_frame.viewmodel

import android.app.Application
import androidx.lifecycle.MutableLiveData
import com.blankj.utilcode.util.ToastUtils
import com.hjc.library_common.viewmodel.KotlinViewModel
import com.hjc.module_frame.entity.ArticleBean
import com.hjc.module_frame.model.LoadSirModel
import java.net.SocketTimeoutException
import java.net.UnknownHostException

class LoadSirViewModel(application: Application) : KotlinViewModel(application) {

    private val mModel = LoadSirModel()

    val listData = MutableLiveData<MutableList<ArticleBean>>()
    val refreshData = MutableLiveData<Boolean>()

    fun loadArticleList(page: Int, isShowProgress: Boolean) {
        launchWrapper({
            mModel.getArticleList(page)
        }, { result ->
            refreshData.value = true

            val data = result?.datas
            data?.let {
                if (it.size > 0) {
                    showContent()
                    listData.value = it
                } else {
                    if (page == 0) {
                        showEmpty()
                    } else {
                        showContent()
                        ToastUtils.showShort("没有更多数据了")
                    }
                }
            }
        }, { e ->
            refreshData.value = true

            if (e is UnknownHostException || e is SocketTimeoutException) {
                showError()
            } else {
                showTimeout()
            }
        }, isShowProgress = isShowProgress)
    }

}